src/main/kotlin/org/rust/ide/inspections/RsVariableMutableInspection.kt

```diff
@@ -0,0 +1,53 @@
+package org.rust.ide.inspections
+
+import com.intellij.codeInspection.ProblemsHolder
+import com.intellij.psi.PsiElement
+import com.intellij.psi.search.LocalSearchScope
+import com.intellij.psi.search.searches.ReferencesSearch
+import org.rust.ide.annotator.fixes.RemoveMutableFix
+import org.rust.lang.core.psi.*
+import org.rust.lang.core.psi.ext.*
+import org.rust.lang.core.resolve.ref.RsMethodCallReferenceImpl
+
+class RsVariableMutableInspection : RsLocalInspectionTool() {
+    override fun getDisplayName() = "No mutable required"
+    override fun buildVisitor(holder: ProblemsHolder, isOnTheFly: Boolean) =
+        object : RsVisitor() {
+            override fun visitPatBinding(o: RsPatBinding) {
+                if (!o.isMut) return
+                val block = o.parentOfType<RsBlock>() ?: o.parentOfType<RsFunction>() ?: return
+                val macro = block.descendantsOfType<RsMacroExpr>()
+                if (macro.any { checkExprPosition(o, it) }) return
+                if (ReferencesSearch.search(o, LocalSearchScope(block))
+                    .asSequence()
+                    .any { checkOccurrenceNeedMutable(it.element.parent) }) return
+                holder.registerProblem(o, "Variable ${o.identifier.text} does not need to be mutable", *arrayOf(RemoveMutableFix(o)))
```

Comment:
There should be backquotes around the variable id in the message. Markdown formatting in my comment just hid them :)